### ISSUE TITLE
rosbag: allow writing message to rosbag without knowing its type (C++)

### ISSUE THREAD

**Issue by fzimmerm** [ros/ros_comm#1417]
Hi,

I'm using the C++ rosbag API, and want to delete a single message from a bag. For this, I was thinking of opening the bag in `Read` mode, opening another bag in`Write` mode, and then iterate over all `MessageInstance` in a `View` of the original bag and write them to the new bag, except for messages with the same topic, type and time as the message I want to delete. But since the bag may contain messages of unknown types, I cannot instantiate the `MessageInstance`. I saw that `MessageInstance` has a method for writing it into a `Stream`, and that `Bag` has a private method `void Bag::write(char const* s, std::streamsize n)`, so in theory it should be possible to serialize all `MessageInstance` into a `Stream` and writing it into the new bag, without knowing the individual message types, if the `write` method weren't private.

Long story short, it would be very useful to have a method to transfer messages between bags without knowing each message's type, in C++ (I think in python this doesn't pose a problem).

**Comment by ibtaylor** [comment]
`auto msg = message_instance.instantiate<topic_tools::ShapeShifter>()`
You should be able to then write that to another bag without having the c++ message type.

Unfortunately, the message instance only has when the message was written "bag write time" and not the "message time". For that, if you want to handle arbitrary types, you might want to look into something like `ros_type_introspection`.

**Comment by fzimmerm** [comment]
Thanks a lot for the hint, this should solve my problem!
I'll post more feedback as soon as I have the time to try it out.

**Comment by fzimmerm** [comment]
I've tried writing all messages to a new bag using `topic_tools::ShapeShifter`. The resulting bag looks fine (same size, `rosbag info` output looks correct), but when iterating over it using a `rosbag::View` it will suddenly throw a `ros::serialization::StreamOverrunException`. This isn't the case with the original bag. The same thing happened when using python to rewrite the bag.

**Comment by cwecht** [comment]
@fzimmerm could you provide a minimal example of your code and a (ideally small) rosbag which reproduce this issue?

**Comment by karthiknit** [comment]
@fzimmerm Could you post a sample code here for writing all messages to a bag using topic_tools::ShapeShifter?

**Comment by fzimmerm** [comment]
@cwecht @karthiknit
Sorry but since I created this issue I've changed position, don't have access to the code in question anymore and am working on a topic unrelated to ROS. Thank you anyway!

**Comment by shaozu** [comment]
```
for(rosbag::MessageInstance const mi : view)
{
    out_bag.write(mi.getTopic(), mi.getTime(), mi);
}
```
Just found this will do the job.

**Comment by karthiknit** [comment]
I think this does the job only when you are reading from one bag and writting data to other bag

**Comment by shaozu** [comment]
Yes, and I guess this will meet the requirement of **fzimmerm**.

**Comment by fjandrad** [comment]
I can confirm that [solution by @shaozu ](https://github.com/ros/ros_comm/issues/1417#issuecomment-819561371)  works. I am able to ready a bag separate topics between the ones I want the ones I dont want and then write to the new bag only the ones I desire.

**Comment by sloretz** [comment]
ROS 1 is end-of-life (EOL) as of today, May 31st 2025. I am archiving this repository because:

* it only supports ROS 1
* it isn't needed anymore in ROS 2

[If you still rely on ROS 1, read this page to learn about your options.](https://bit.ly/NoeticEOL)
